common/internal_model/src/main/scala/weco/catalogue/internal_model/identifiers/ReferenceNumber.scala

```diff
@@ -0,0 +1,35 @@
+package weco.catalogue.internal_model.identifiers
+
+import io.circe.{Decoder, Encoder, HCursor, Json}
+
+class ReferenceNumber(val underlying: String) {
+  override def toString: String = underlying
+
+  // Normally we use case classes for immutable data, which provide these
+  // methods for us.
+  //
+  // We deliberately don't use case classes here so we skip automatic
+  // case class derivation for JSON encoding/DynamoDB in Scanamo,
```

Comment:
Could you explain why this is useful?

Reply:
This is a pattern copied from a bunch of other places where we have types that are fancy names for strings – e.g. CanonicalId, SierraRecordNumber, ExternalIdentifier in the storage service. We want them to be distinct types in Scala so you don't mix them up.

If you used a case class, you'd get a JSON serialisation something like:

```json
{"underlying": "12345i"}
```

whereas our custom encoder/decoder will turn it into a string:

```
"12345i"
```

And the latter is preferable because it's a bit easier for any other tools working with the data (one less layer of indirection), and it's backwards-compatible with any existing data stores that used the old string-based serialisation.

Reply:
Ah, OK I see - so to avoid unintentional automatic derivation, we just don't let it be possible. 

I do feel having the auto stuff is _occasionally_ useful, but also creates compilation performance errors, and creates issues like this, where we have to create a special class for something that already exists.

Should we remove auto-derivation and either expect all case classes to define their encoders, or have a mega-encoder like [this](https://github.com/wellcomecollection/catalogue-pipeline/blob/main/common/internal_model/src/main/scala/uk/ac/wellcome/models/Implicits.scala)?

Reply:
bikeshedding - Not for this PR

Reply:
That mega-encoder is still using the auto-derivation, it’s just caching the results so they don't have to be re-derived every time.

Reply:
Yea, true. I mean that we could use a `derivedDecoder`, but explicitly set it on a case class, in the case of the above, have a special one.

Reply:
I'm going to go ahead and extract some of this logic in a separate PR, because we repeat it quite a lot – and then we can add a thorough set of comments and tests around it.

Reply:
Sounds great!

Reply:
✨ https://github.com/wellcomecollection/scala-libs/pull/98 ✨